Add spec for SPV_NV_shader_invocation_reorder. (#175)
KhronosGroup/SPIRV-Registry: extensions/NV/SPV_NV_shader_invocation_reorder.asciidoc

OpHitObjectRecordHitMotionNV
OpHitObjectRecordEmptyNV
OpHitObjectTraceRayNV
OpHitObjectTraceRayMotionNV
----

Modifications to the SPIR-V Specification
-----------------------------------------

(Modify Section 2.2.1, Instructions ) ::
+
_Shader Call Instruction_: An instruction which may cause execution to
continue elsewhere by creating one or more invocations that execute
other shaders. The <<OpTraceRayWithHitObjectNV,*OpTraceRayWithHitObjectNV*>>,
<<OpHitObjectExecuteShaderNV,*OpHitObjectExecuteShaderNV*>> instructions are
shader call instructions. +
+
_Invocation Repack Instruction_: An instruction which is a shader call where the
implementation may repack shader invocations, before and/or after execution
 of the called shader. The <<OpReorderThreadWithHintNV,*OpReorderThreadWithHintNV*>>,
 and <<OpReorderThreadWithHitObjectNV,*OpReorderThreadWithHitObjectNV*>> are invocation
 repack_instructions.

(Add the following terminology to section 2.2.2, Types) ::
+
--
'Hit object type': The type returned by *OpTypeHitObjectNV*.
--

(Add to the list of opaque types in section 2.2.2, Types) ::
+
--
 * *OpTypeHitObjectNV*
--

(Modify Section 2.16.1, Universal Validation Rules) ::

> --

Reviewer: Missing edits for this section? (or remove the header)

(Modify Section 3.2, Decorations, adding a row to the Decoration table) ::
+
--
[cols="1^.^,25,15",options="header",width = "100%"]
:capability: ShaderInvocationReorderNV
|====
2+^.^| Decoration | Requires | *ShaderInvocationReorderNV*
| 5386 | *HitObjectShaderRecordBufferNV* +
|====
--

(Modify Section 3.31, Capability, adding a row to the Capability table) ::
+
--
[cols="1^.^,25,15",options="header",width = "100%"]
:capability: ShaderInvocationReorderNV
|====
2+^.^| Capability | Implicitly Declares
| 5383 | *ShaderInvocationReorderNV* +
| *Shader*
|====
--

> (Modify Section 3.36.6, Type-Declaration Instructions, adding two new tables) ::

Reviewer: ```suggestion
(Modify Section 3.36.6, Type-Declaration Instructions, adding a new table) ::
```

Reviewer: Edits to section 3.36 should be after edits to section 3.32 below.

+
--
:capability: ShaderInvocationReorderNV

[cols="5,1,1*3",width="100%"]
|=====
2+|[[OpTypeHitObjectNV]]*OpTypeHitObjectNV* +
 +
Declares a hit object type which is an opaque object representing state during
ray tracing traversal.

This type is opaque: values of this type have no defined physical size or
bit pattern.

1+|Capability: +
*ShaderInvocationReorderNV*
| 2 | 5281 | '<id>' 'Result'
|=====
--

(Add the following line to the description of OpTypePointer, in Section 3.32.6, Type-Declaration Instructions) ::
+
--
If 'Type' is *OpTypeHitObjectNV*, 'Storage Class' must be 'Private' or 'Function'.
--

(Add the following line to the description of OpStore and OpLoad, in Section 3.32.8, Memory Instructions) ::
+
--
The 'Type' operand to the *OpTypePointer* used for 'Pointer' must not be *OpTypeHitObjectNV*.
--

(Add the following line to the description of OpCopyMemory and OpCopyMemorySized, in Section 3.32.8, Memory Instructions) ::
+
--
The 'Type' operand to the *OpTypePointer* used for 'Target' or 'Source' must not be *OpTypeHitObjectNV*.
--

> (Modify Section 3.7, Storage Class, adding rows to the Storage Class table) ::

Reviewer: edits to section 3.7 should be earlier (please keep sections in order)

+
--
[cols="1^,15,2*2",options="header",width = "100%"]
|====
2+^.^| Storage Class | <<Capability,Enabling Capabilities>> | Enabled by Extension
| 5385 | *HitObjectAttributeNV* +
Used for storing attributes of geometry intersected by a ray to be passed on to
hit object instructions. Visible across all functions in the current invocation. 
Not shared externally. Variables declared with this storage class can be both read and written to, but cannot have initializers.
Only allowed in *RayGenerationKHR*, *ClosestHitKHR*, and *MissKHR* execution models.
| *ShaderInvocationReorderNV* | SPV_NV_shader_invocation_reorder
|====
--

(Add a new sub section 3.36.Reorder Instructions) ::
+
--
[cols="4*1"]
|======
3+|[[OpReorderThreadWithHintNV]]*OpReorderThreadWithHintNV* +
 +

>  Reorder threads based on user provided hint. Similar hint values indicate similarity of     subsequent work done after this call. Behavior is implementation defined +

Reviewer: ```suggestion
 Reorder threads based on user provided hint. Similar hint values indicate similarity of     subsequent work done after this call. Behavior is implementation-defined. +
```

 +
 'Hint' must be a 32-bit 'integer-type' scalar. +
 +
 'Bits' must be a 32-bit 'integer-type' scalar. +
 +
 This instruction is allowed only in *RayGenerationKHR* execution models. +
1+|Capability: +
*ShaderInvocationReorderNV*
| 3 | 5280
| '<id> Hint'
| '<id> Bits'
|======

[cols="5*1"]
|======
4+|[[OpReorderThreadWithHintNV]]*OpReorderThreadWithHitObjectNV* +
 +

>  Reorder threads based on hit object supplemented by hint and bits if they are optionally    provided values. Behavior is implementation defined +

Reviewer: ```suggestion
 Reorder threads based on hit object supplemented by 'Hint' and 'Bits' if they are optionally provided. Behavior is implementation-defined. +
```

| '<id>' 'Hit Object'
|======

[cols="5*1"]
|======
4+|[[OpHitObjectIsMissNV]]*OpHitObjectIsMissNV* +
 +
 Returns a boolean indicating whether the hit object has encoded a miss. +
 +
 'Result' is true if hit object encodes a miss, false otherwise. +
 +
 'Result Type' must be a 'boolean type' scalar. +
 +
 'Hit Object' must be a pointer to hit object. +
 +
 This instruction is allowed only in *RayGenerationKHR*, *ClosestHitKHR* and *MissKHR* execution models. +
1+|Capability: +
*ShaderInvocationReorderNV*
| 3 | 5278
| '<id>' 'Result Type'
| '<id>' 'Result'
| '<id>' 'Hit Object'
|======
--

--
Validation Rules
----------------

An OpExtension must be added to the SPIR-V for validation layers to check
legal use of this extension:

----
OpExtension "SPV_NV_shader_invocation_reorder"
----

Interactions with SPV_KHR_ray_tracing
-------------------------------------

This extension requires SPV_KHR_ray_tracing

Issues
------

None

Revision History
----------------

[cols="5,15,15,70"]
[grid="rows"]
[options="header"]
|========================================
|Rev|Date|Author|Changes
|1 |2022-08-01 |Ashwin Lele         | Internal revisions
|========================================

> 

Reviewer: Please build the asciidoc and include the .html in the PR

Link the html from the top-level README.md

Reviewer: added in lasted pushes.
